src/rgw/driver/rados/rgw_reshard.cc

```diff
@@ -236,16 +237,49 @@ class BucketReshardShard {
     return 0;
   }
 
-  int flush() {
+  int flush(bool process_log = false, RGWBucketReshard *br = nullptr,
+            const DoutPrefixProvider *dpp = nullptr) {
     if (entries.size() == 0) {
       return 0;
     }
 
     librados::ObjectWriteOperation op;
-    for (auto& entry : entries) {
-      store->getRados()->bi_put(op, bs, entry, null_yield);
+    if (process_log) {
+      map<RGWObjCategory, rgw_bucket_category_stats> dec_stats;
+      list<rgw_cls_bi_entry> dec_entries;
+      set<string> dec_entry_names_wanted;
+      for (auto& entry : entries) {
+        store->getRados()->bi_put(op, bs, entry, null_yield);
+        dec_entry_names_wanted.emplace(entry.idx);
+      }
+
+      // getting the index entry in target shard
+      int ret = store->getRados()->bi_get_vals(bs, dec_entry_names_wanted, &dec_entries, null_yield);
```

Comment:
> i guess the alternative would be to teach rgw_bi_put_op() in cls_rgw.cc to look for an existing key

I remembered using this way at first, but gave up at last. I will research again and do it if can.

Reply:
yeah, there's a tradeoff to consider. doing this in `rgw_bi_put_op()` adds an extra omap read to each call, and an extra read-header+write-header to each call that finds an existing key

doing this in `flush()` allows us to batch these stat updates together with `cls_rgw_bucket_update_stats()`, at the cost of an extra round trip for `bi_get_vals()` that blocks `reshard_process()`

Reply:
> rgw_bi_put_op() adds an extra omap read to each call, and an extra read-header+write-header to each call that finds an existing key

@cbodley  I Implemented in this way, and find the reason again. A batch `bi_put()` are transferred together. They will be put in a transaction after `do_osd_ops()`, this will make read-header with old stats.

Reply:
ok, makes sense. will resolve this for now

Reply:
to work around this issue with read-modify-writes over several `bi_put()` calls, we could introduce a new `bi_put_entries()` call that takes the entire vector of entries and updates all of their stats at once

that avoids the need for this separate `bi_get_vals()` api and, most importantly, allow `flush()` to be fully-async again

i opened a draft https://github.com/ceph/ceph/pull/59581 on top of your pr with some additional commits for this